README.md
# proxy-registry

This is a simple proxy registry that uses your existing npm cache.

```console
$ npx proxy-registry
Listening on: https://localhost:22000
To use: npm config set registry https://localhost:22000
^C to close server
```

## OPTIONS

```
Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]
  --port     the port to listen on                     [number] [default: 22000]
  --shell    run a shell configured to talk to this proxy
                                                       [boolean] [default: true]
  --log      log requests (defaults to off when running a shell, on when not)
                                                                       [boolean]
```



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.3"></a>
## [1.0.3](https://github.com/npm/proxy-registry/compare/v1.0.2...v1.0.3) (2018-08-07)


### Bug Fixes

* **macos:** localhost -> 127.0.0.1 ([bb63a3f](https://github.com/npm/proxy-registry/commit/bb63a3f))
* **packument:** Switch packument fetching to pacote based code ([f658a48](https://github.com/npm/proxy-registry/commit/f658a48))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/npm/proxy-registry/compare/v1.0.1...v1.0.2) (2018-07-31)



<a name="1.0.1"></a>
## 1.0.1 (2018-07-31)


config.js
const qx = require('@perl/qx')

const conf = exports.npm = JSON.parse(qx.sync`npm config ls -l --json`)
exports.pacote = {
  cache: getCacheMode(conf),
  cacheManager: `${conf.cache}/_cacache`,
  ca: conf.ca,
  cert: conf.cert,
  headers: getHeaders('', conf.registry, conf),
  key: conf.key,
  localAddress: conf['local-address'],
  maxSockets: conf['maxsockets'],
  proxy: conf.proxy,
  retry: conf.retry,
  strictSSL: !!conf['strict-ssl'],
  timeout: conf.timeout,
  uid: conf.uid,
  gid: conf.gid
}

function getCacheMode (opts) {
  return opts.offline
    ? 'only-if-cached'
    : opts['prefer-offline']
      ? 'force-cache'
      : opts['prefer-online']
        ? 'no-cache'
        : 'default'
}
function getHeaders (uri, registry, opts) {
  const headers = Object.assign({
    'user-agent': opts['user-agent']
  }, opts.headers)
  // check for auth settings specific to this registry
  let auth = (
    opts.auth &&
    opts.auth[registryKey(registry)]
  ) || opts.auth
  // If a tarball is hosted on a different place than the manifest, only send
  // credentials on `alwaysAuth`
  const shouldAuth = auth && (
    auth.alwaysAuth ||
    url.parse(uri).host === url.parse(registry).host
  )
  if (shouldAuth && auth.token) {
    headers.authorization = `Bearer ${auth.token}`
  } else if (shouldAuth && auth.username && auth.password) {
    const encoded = Buffer.from(
      `${auth.username}:${auth.password}`, 'utf8'
    ).toString('base64')
    headers.authorization = `Basic ${encoded}`
  } else if (shouldAuth && auth._auth) {
    headers.authorization = `Basic ${auth._auth}`
  }
  return headers
}

function registryKey (registry) {
  const parsed = url.parse(registry)
  const formatted = url.format({
    host: parsed.host,
    pathname: parsed.pathname,
    slashes: parsed.slashes
  })
  return url.resolve(formatted, '.')
}


fetch-packument.js
'use strict'
const fetch = require('npm-registry-fetch')
const optCheck = require('pacote/lib/util/opt-check')
const BB = require('bluebird')
const LRU = require('lru-cache')
const CORGI_DOC = 'application/vnd.npm.install-v1+json; q=1.0, application/json; q=0.8, */*'
const JSON_DOC = 'application/json'

module.exports = packument

function packument (name, opts) {
  opts = optCheck(opts)
  const registry = fetch.pickRegistry(name, opts)
  const uri = registry.replace(/\/?$/, `/${name.replace(/[/]/g, '%2f')}`)
  return fetchPackument(uri, name, registry, opts)
}

function fetchPackument (uri, spec, registry, opts) {
  const mem = pickMem(opts)
  if (mem && !opts.preferOnline && mem.has(uri)) {
    return BB.resolve(mem.get(uri))
  }
  return fetch(uri, opts.concat({
    headers: {
      'pacote-req-type': 'packument',
      'pacote-pkg-id': `registry:${spec}`,
      accept: opts.fullMetadata ? JSON_DOC : CORGI_DOC
    },
    spec
  }, opts, {
    // Force integrity to null: we never check integrity hashes for manifests
    integrity: null
  })).then(res => res.json().then(packument => {
    packument._cached = decodeURIComponent(res.headers.has('x-local-cache'))
    packument._contentLength = +res.headers.get('content-length')
    // NOTE - we need to call pickMem again because proxy
    //        objects get reused!
    const mem = pickMem(opts)
    if (mem) {
      mem.set(uri, packument)
    }
    return packument
  }))
}

const MEMO = new LRU({
  length: m => m._contentLength,
  max: 200 * 1024 * 1024, // 200MB
  maxAge: 30 * 1000 // 30s
})

class ObjProxy {
  get (key) { return this.obj[key] }
  set (key, val) { this.obj[key] = val }
}

// This object is used synchronously and immediately, so
// we can safely reuse it instead of consing up new ones
const PROX = new ObjProxy()
function pickMem (opts) {
  if (!opts || !opts.memoize) {
    return MEMO
  } else if (opts.memoize.get && opts.memoize.set) {
    return opts.memoize
  } else if (typeof opts.memoize === 'object') {
    PROX.obj = opts.memoize
    return PROX
  } else {
    return null
  }
}


server.js
#!/usr/bin/env node
'use strict'
require('@iarna/cli')(main)
  .option('port', {
    describe: 'the port to listen on',
    type: 'number',
    default: 22000
  })
  .option('web-port', {
    describe: 'the port to listen on',
    type: 'number',
    default: 22080
  })
  .option('shell', {
    describe: 'run a shell configured to talk to this proxy',
    type: 'boolean',
    default: true
  })
  .option('log', {
    describe: 'log requests (defaults to off when running a shell, on when not)',
    type: 'boolean'
  })

const fun = require('funstream')
const ssri = require('ssri')
const spawn = require('child_process').spawn
const pacote = require('pacote')
const fetch = require('make-fetch-happen')
const qr = require('@perl/qr')
const http = require('http')
const cache = require('cacache/en')
const Koa = require('koa')
const compress = require('koa-compress')
const logger = require('koa-logger')
const alwaysJson = require('./always-json.js')
const fetchPackument = require('./fetch-packument.js')
const tar = require('tar')
const semver = require('semver')
const conf = require('./config.js')
let marky
try {
  marky = require("marky-markdown")
} catch (_) {
  const MarkdownIt = require('markdown-it')
  marky = (_, opts) => {
    const md = new MarkdownIt({
      html: true,
      linkify: true,
      ...opts
    })
    return md.render(_)
  }
}

const registry = conf.npm.registry.replace(/[/]$/, '')

const cacheDir = conf.npm.cache + '/_cacache'

async function main (opts, ...args) {
  if (opts.log == null) opts.log = !opts.shell

  const reg = new Koa()
  if (opts.log) reg.use(logger())
  reg.use(alwaysJson())
  reg.use(compress())
  reg.use(registryRequest)
  const regHttp = http.createServer(reg.callback()).listen(opts.port)

  const web = new Koa()
  if (opts.log) web.use(logger())
  web.use(compress())
  web.use(webRequest)
  const webHttp = http.createServer(web.callback()).listen(opts['web-port'])

  await Promise.race([new Promise((resolve, reject) => {
    reg.on('error', reject)
    regHttp.on('error', reject)
    if (opts.shell) {
      process.env['npm_config_registry'] = `http://127.0.0.1:${opts.port}`
      console.log(`Starting subshell configured to talk to: http://127.0.0.1:${opts.port}`)
      console.log(`Web server: http://127.0.0.1:${opts['web-port']}`)
      console.log(`To close server, run: exit`)
      spawn(conf.npm.shell, [], {stdio: 'inherit'})
        .on('close', er => er ? reject(er) : resolve())
    } else {
      console.log(`Web server: http://127.0.0.1:${opts['web-port']}`)
      console.log(`To use: npm config set registry http://127.0.0.1:${opts.port}`)
      console.log(`^C to close server`)
      process.on('SIGINT', resolve)
    }
  }), new Promise((resolve, reject) => {
    web.on('error', reject)
    webHttp.on('error', reject)
  })])
  console.error('\nShutting down')
  regHttp.close()
  webHttp.close()
}

const matchName = qr`(?:@[^/]+/)?[^/]+`
const matchVersion = qr`\d+\.\d+\.\d+(?:-.*)?`

const matchWebPackage = qr`^/package/(${matchName})(?:/(${matchVersion}))?(?:[?].*)?$`
async function webRequest (ctx, next) {
  try {
    if (ctx.request.url === '/') {
      await webHome(ctx)
    } else if (matchWebPackage.test(ctx.request.url)) {
      const [, name, version] = matchWebPackage.exec(ctx.request.url)
      await webPackage(ctx, name, version)
    } else {
console.error(matchWebPackage, ctx.request.url)
      await webNotFound(ctx)
    }
  } catch (ex) {
    console.error(ex)
    ctx.response.status = 500
    ctx.response.body = `<pre>${JSON.stringify(ex)}</pre>`
  }
  await next()
}

async function webNotFound (ctx) {
  ctx.response.status = 404
  ctx.response.body = '<h1>Not found</h1>'
}

async function tarballMetadata (tarball) {
  const tb = cache.get.stream(cacheDir, `make-fetch-happen:request-cache:${tarball}`)
  let readmeP
  let pjsonP
  let shrinkP
  await fun(tb).pipe(tar.t()).on('entry', async entry => {
    if (!readmeP && qr.i`^[^/]+/readme(?:$|[.])`.test(entry.path)) readmeP = entry.pipe(fun()).concat()
    if (!pjsonP && qr`^[^/]+/package.json`.test(entry.path)) pjsonP = entry.pipe(fun()).concat()
    if (!shrinkP && qr`^[^/]+/npm-shrinkwrap.json`.test(entry.path)) shrinkP = entry.pipe(fun()).concat()
  })
  const [readme, pjson, shrink] = await Promise.all([readmeP, pjsonP, shrinkP])
  return {readme, manifest: pjson && JSON.parse(pjson), shrinkwrap: shrink && JSON.parse(shrink)}
}

async function webPackage (ctx, name, version) {
  const ent = await cache.get(cacheDir, `make-fetch-happen:request-cache:${registry}/${name.replace('/', '%2f')}`)
  const packument = JSON.parse(ent.data)
  if (!version) {
    version = packument['dist-tags'].latest
  }
  let manifest = packument.versions[version]
  const tarball = manifest.dist.tarball
  const integrity = manifest.dist.integrity || ssri.fromHex(manifest.dist.shasum, 'sha1').toString()
  let readme = ''
  try {
    const pkg = await tarballMetadata(tarball)
    if (pkg.readme) readme = pkg.readme
    if (pkg.manifest) manifest = pkg.manifest
  } catch (_) {
  }
  ctx.response.body = `<html>
<head>
   <link rel="stylesheet" href="https://static.npmjs.com/862d62e1eccfac06e291a1603a2ec56c.css"/>
   <link rel="stylesheet" href="https://static.npmjs.com/cms/flatpages.css"/>
   <link rel="stylesheet" href="https://static.npmjs.com/6f0fec69a6599ac07cbe906fef441123.css"/>
  <style>
   body { margin: 5em; }
    #readme { float: left; width: 75%; }
    .manifest { width: 23%; overflow: none; }
  </style>
</head>
<body>
  <div id="readme">${marky(readme, {package: manifest})}</div>
  <div class="manifest"><pre>${JSON.stringify(manifest, null, 2)}</pre></div>
</body>
</html>
`
  ctx.response.status = 200
}

async function webHome (ctx) {
  ctx.response.body = `<html>
  <body>
    ${await listModules()}
  </body>
</html>`
  ctx.response.status = 200

}

const matchCacheTarball = qr`^make-fetch-happen:request-cache:(https?://[^/]+/(${matchName})/-/.*?-(${matchVersion})[.]tgz)$`

const matchCachePackument = qr`^make-fetch-happen:request-cache:https?://[^/]+/(${matchName})$`
async function listModules() {
  const modules = {}
  Object.values(await cache.ls(cacheDir)).forEach(_ => {
    if (matchCacheTarball.test(_.key)) {
      const [, tarball, name, version] = matchCacheTarball.exec(_.key)
      if (!modules[name]) modules[name] = {name, versions: {}}
      if (!modules[name].tarball) modules[name].tarball = tarball
      modules[name].versions[version] = {name, version, tarball}
    } else if (matchCachePackument.test(_.key)) {
      const [, name] = matchCachePackument.exec(_.key)
      if (!modules[name]) modules[name] = {name, versions: {}}
      modules[name].packument = true
    }
  })
  const listing = Object.values(modules).filter(_ => _.tarball && _.packument).sort((aa, bb) => aa.name.localeCompare(bb.name))
  return fun(listing).map(async _ => {
    try {
      const pkg = await tarballMetadata(_.tarball)
      return `<b><a href="/package/${_.name}">${_.name}</a></b> (` +
        Object.values(_.versions).sort((aa, bb) => -1 * semver.compare(aa.version, bb.version)).map(_ => `<a href="/package/${_.name}/${_.version}">${_.version}</a>`).join(', ') +
        `)${pkg.manifest.description ? ': ' + pkg.manifest.description : ''}`
    } catch (_) {
      console.log(_)
    }
  }).filter(_ => _).grab(_ => _.join('<br>\n'))
}

const matchTarball = qr`^/(${matchName})/-/.*?-(${matchVersion})\.tgz$`
const matchManifest = qr`^/(${matchName})$`

async function registryRequest (ctx, next) {
  const requestConfig = Object.assign({}, conf.pacote, {headers: ctx.request.header})
  delete requestConfig.headers.host
  requestConfig.method = ctx.request.method

  try {
    if (matchTarball.test(ctx.request.url)) {
      await fetchTarball(ctx, requestConfig)
    } else if (matchManifest.test(ctx.request.url)) {
      await fetchManifest(ctx, requestConfig)
    } else {
      const result = await proxyRequest(ctx.request.url, ctx, requestConfig)
      ctx.response.body = result.body
    }
  } catch (ex) {
    console.error(ex)
    ctx.response.status = 500
    ctx.response.body = JSON.stringify(ex)
  }
  await next()
}

function fetchTarball (ctx, requestConfig) {
  const [, name, version] = matchTarball.exec(ctx.request.url)
  ctx.response.body = pacote.tarball.stream(`${name}@${version}`, requestConfig)
}

async function fetchManifest (ctx, requestConfig) {
  const [, name] = matchManifest.exec(ctx.request.url)
  const body = await fetchPackument(name, requestConfig)
  for (let version of Object.keys(body.versions)) {
    let vv = body.versions[version]
    vv.dist.tarball = vv.dist.tarball.replace(qr.g`${registry}`, `http://127.0.0.1:22000`)
  }
  ctx.response.body = JSON.stringify(body)
}

async function proxyRequest (url, ctx, requestConfig) {
  if (requestConfig.method === 'PUT' || requestConfig.method === 'POST') requestConfig.body = ctx.req

  const result = await fetch(`${registry}${url}`, requestConfig)
  for (let header of result.headers.entries()) {
    const [key, value] = header
    if (key === 'transfer-encoding' || key === 'content-encoding' || key === 'content-length' || key === 'connection') continue
    ctx.response.set(key, value)
  }
  ctx.response.status = result.status
  return result
}
